File: task4.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int determineWinner(int n, int m, vector<int> a) {

    int pavel_score = 0;
    int vika_score = 0;

    int pavel_last = -1;
    int vika_last = -1;

    bool pavel_turn = true;

    while (!a.empty()) {

        int best_k = -1;
        int best_sum = -1e9;

        int current_sum = 0;

        for (int k = 1; k <= m && k <= a.size(); k++) {

            current_sum += a[k - 1];

            // проверка ограничения
            if (pavel_turn && k == pavel_last) continue;
            if (!pavel_turn && k == vika_last) continue;

            // выбираем максимум, при равенстве — минимальный k
            if (current_sum > best_sum) {
                best_sum = current_sum;
                best_k = k;
            }
        }

        // начисляем очки
        if (pavel_turn) {
            pavel_score += best_sum;
            pavel_last = best_k;
        } else {
            vika_score += best_sum;
            vika_last = best_k;
        }

        // удаляем элементы
        a.erase(a.begin(), a.begin() + best_k);

        // смена хода
        pavel_turn = !pavel_turn;
    }

    return pavel_score > vika_score ? 1 : 0;
}

int main() {

    int n, m;
    cout << "Введите n и m: ";
    cin >> n >> m;

    vector<int> numbers(n);

    cout << "Введите числа: ";
    for (int i = 0; i < n; i++) {
        cin >> numbers[i];
    }

    int result = determineWinner(n, m, numbers);

    cout << "Результат: " << result << endl;

    return 0;
}

File: task4.go
package main

import (
	"fmt"
)

// функция определения победителя
func determineWinner(n int, m int, a []int) int {

	pavelScore := 0
	vikaScore := 0

	pavelLast := -1
	vikaLast := -1

	pavelTurn := true

	// пока есть числа
	for len(a) > 0 {

		bestK := -1
		bestSum := -1000000000

		currentSum := 0

		// перебираем сколько элементов взять
		for k := 1; k <= m && k <= len(a); k++ {

			currentSum += a[k-1]

			// проверка ограничения варианта 1
			if pavelTurn && k == pavelLast {
				continue
			}
			if !pavelTurn && k == vikaLast {
				continue
			}

			// выбираем максимум, при равенстве — минимальный k
			if currentSum > bestSum {
				bestSum = currentSum
				bestK = k
			}
		}

		// начисление очков
		if pavelTurn {
			pavelScore += bestSum
			pavelLast = bestK
		} else {
			vikaScore += bestSum
			vikaLast = bestK
		}

		// удаляем выбранные элементы
		a = a[bestK:]

		// смена хода
		pavelTurn = !pavelTurn
	}

	// определяем победителя
	if pavelScore > vikaScore {
		return 1
	}
	return 0
}

func main() {

	var n, m int

	fmt.Print("Введите n и m: ")
	fmt.Scan(&n, &m)

	numbers := make([]int, n)

	fmt.Print("Введите числа: ")
	for i := 0; i < n; i++ {
		fmt.Scan(&numbers[i])
	}

	result := determineWinner(n, m, numbers)

	fmt.Println("Результат:", result)
}
